File: Queen.py
from Piece import Piece
from Cell import Cell


class Queen(Piece):
    def _slide(self, board, dr, dc, row, col):
        if 0 <= row + dr <= 7 and 0 <= col + dc <= 7:
            if board[row + dr][col + dc] is None:
                self.targets.append(Cell(row + dr, col + dc))
                self._slide(board, dr, dc, row + dr, col + dc)
            elif board[row + dr][col + dc].team != self.team:
                self.targets.append(Cell(row + dr, col + dc))

    def calc_targets(self, board):
        self.targets = []
        self._slide(board, 1, 1, self.row, self.col)
        self._slide(board, -1, 1, self.row, self.col)
        self._slide(board, 1, -1, self.row, self.col)
        self._slide(board, -1, -1, self.row, self.col)
        self._slide(board, 1, 0, self.row, self.col)
        self._slide(board, -1, 0, self.row, self.col)
        self._slide(board, 0, 1, self.row, self.col)
        self._slide(board, 0, -1, self.row, self.col)
        if self.critical:
            super().filter_to_pin_ray()

    def get_value(self, board):
        total = 49
        self.calc_targets(board)
        for cell in self.targets:
            total += 1
            if (cell.row == 3 or cell.row == 4) and (cell.col == 3 or cell.col == 4):
                total += 1
        return total

    def print_piece(self):
        print("Queen at", self.row, ",", self.col)

File: Chess101iOS/Sources/Chess101Engine/Queen.swift
public final class Queen: Piece {
    public var row: Int
    public var col: Int
    public var team: Team
    public var targets: [Cell] = []
    public var touched: Bool = false
    public var critical: Bool = false
    public var criticalTargets: [Cell] = []
    public var typeName: String { "Queen" }

    public init(row: Int, col: Int, team: Team) {
        self.row = row; self.col = col; self.team = team
    }

    @discardableResult
    public func calcTargets(board: Board) -> Bool? {
        targets = []
        let dirs = [(0,1),(0,-1),(1,0),(-1,0),(1,1),(1,-1),(-1,1),(-1,-1)]
        for (dr, dc) in dirs { bladeRunner(board: board, dr: dr, dc: dc, fromRow: row, fromCol: col) }
        if critical { criticalMan() }
        return nil
    }

    public func move(toRow: Int, toCol: Int, board: Board) -> MoveResult {
        row = toRow; col = toCol; touched = true
        return .normal
    }

    public func getValue(board: Board) -> Int {
        var val = 49
        val += targets.count
        return val
    }
}
